Match Ergonomics 2024: document and reorganize the currently-implemented feature gates

dianne opened the pull request:
The hope here is to make it easier to adjust, understand, and test the experimental pattern typing rules implemented in the compiler. This PR doesn't (or at isn't intended to) change any behavior or add any new tests; I'll be handling that later. I've also included some reasoning/commentary on the more involved changes in the commit messages.

Relevant tracking issue: #123076

r? @Nadrieril

dianne committed:
make experimental pattern typing features mutually exclusive

This aims to reduce the complexity needed in the boolean logic for telling which
rules we're using to type patterns. If we still want the functionality this
removes, we can re-add it later, after some cleanup to pattern typing.

dianne committed:
cleanup: de-tangle experimental pattern typing rules some

The goal of this cleanup is to make it more apparent which feature gates
correspond to which typing rules, and which typing rules correspond to
what code. My intent is for calls to the "which typing rules do we
have?" functions to be replaced by comments (and edition checks, as
appropriate), but as long as we're experimenting with multiple rulesets,
this seemed to me to be the easiest to document and read. There's still
some nontrivial control flow, but I've added comments to try and make it
clearer.

There's some logic that looks like it could be de-duplicated across
different ways of matching against inherited references; however, the
duplication is intentional. Once we choose which rulesets we want, we
can make this more clever, but until then, my priorities are clarity and
ease of modification/extension. That said, I think the diagnostics could
use some work; factoring out commonalities there (and separating them
from the typing logic) would be ideal. I've opted not to include that
here both since it'd make this refactor less obvious and since it
affects test output.

Also, this doesn't get quite as fine-grained as Typing Rust Patterns, so
there's some instances where certain rules are conflated. I'd prefer to
minimize dead/untested codepaths for rulesets we're not interested in,
so as a compromise I've added comments wherever some aspect of the
typing rules is assumed from another. I'm not totally happy with it, but
I think it's at least better than plain checks against the feature gates
and edition.

dianne committed:
move the experimental match ergonomics tests to be with the other rfc 3627 tests

dianne committed:
reorganize and comment some of the experimental pattern typing tests

This only includes previously existing tests (with a couple duplicates
removed). I plan on adding more comprarisons where the rules differ
once I've updated the pattern typing rules. I also haven't touched the
tests for new rules in old editions; I'll see how best to handle that
once those rules are updated as well.

dianne committed:
pattern typing cleanup: remove a redundant assignment

As far as I can tell, the assignment removed here will never do
anything. `pat_info.max_ref_mutbl` starts at `MutblCap::Mut` for the
top-level pattern and is only changed if feature gates are enabled that
would result in the statement not being executed. Regardless of what new
pattern typing rules are adopted, I don't imagine we want to
conditionally reset `max_ref_mutbl` based on edition either, since it'd
have consequences for subpatterns in other editions.

dianne committed:
always track patterns' `MutblCap`

If Rules 3 or 5 are adopted in any edition, we'll need to track the
`MutblCap` in all editions for macro hygiene purposes. Previously, the
check for whether to track it was conflated with the checks for whether
to apply Rules 3 and 5, so to make it a bit clearer, this always tracks
the `MutblCap`. If needed, we could check if Rules 3 or 5 are present in
any edition before tracking the `MutblCap`, but since it's not that much
more expensive to always track it, I've figured that's simplest.

My main concern with removing the checks is that it may not be clear
that the `MutblCap` is tracked for those specific purposes. To try and
mitigate this, I've made its doc comment a bit more precise regarding
the extent of how and why it's used.

This leaves the condition untouched on the `cap_to_weakly_not` call
needed for Rule 5, since it's only needed for that and it can affect
diagnostics.

dianne committed:
rename and comment the test for "Rule 5"-related mutability errors

This also makes it test the "structural" ruleset, in preparation for
additional tests where the rulesets disagree.

dianne committed:
remove `tests/ui/pattern/no_ref_mut_behind_and.rs`

This was a test for `ref_pat_eat_one_layer_2024` downgrading a `mut ref`
default binding mode to `ref` within a shared reference pattern (i.e.
Rule 3) on edition 2021 specifically. Since it's near-identical to
another existing test (currently in `ref_pat_eat_one_layer_2021.rs` in
the "experimental" rfc 3627 subdirectory) and that particular feature
gate's typing rulesets are planned to no longer have Rule 3, I'm opting
to remove it rather than continue maintaining it separately.

dianne commented:
I've added bit more test maintenance. The remaining tests I haven't touched at this point are:
- `feature-gate-ref_pat_eat_one_layer_2024.rs`: this seems to follow the pattern of other gate-tests.
- `ref_pat_eat_one_layer_2021.rs` and `ref_pat_eat_one_layer_2021_fail.rs`: I'm thinking it'll most likely be cleanest to test edition 2021 and 2024 behavior together (at least in most cases), at which point these can be deleted. Until then, I'm leaving them.
- `tests/ui/pattern/mut_preserve_binding_mode_2021.rs` and `tests/ui/pattern/mut_preserve_binding_mode_2024.rs`: these are for the `mut_ref` feature and mirror the `mut-ref-mut.rs` failure case. I'm not sure what the status of `mut_ref` is, but as long as I don't break them, I don't mind leaving them for whoever's maintaining that feature.
- `tests/ui/pattern/skipped-ref-pats-issue-125058.rs`: this is a regression test for an ICE. I'm not sure whether it belongs with the tests that the feature gates behave according to spec, so I've left it for now, though it may want a new home (or at least a comment).

Nadrieril commented:
> I'm not sure what the status of `mut_ref` is

It was made as part of the match ergonomics 2024 work, it's not really maintained nor pushed forward by anyone. Let's keep it around until we're sure we won't need it.

Nadrieril reviewed (approved):
Thanks for this! I'm a fan of making the rules more explicit and separate from the feature gates, this is a great legibility improvement. Much appreciate the small commits also.

Very good point on needing to track `MutblCap` in all editions! Could you add a test for that? We already have a cross-edition test in `rfc-3627-match-ergonomics-2024/migration-lint.rs`, it would make sense to add a `mixed_editions.rs` file with a test or two. Feel free to leave that for a later PR.

Code looks good, I only have suggestions regarding comments.

dianne commented on compiler/rustc_hir_typeck/src/pat.rs:
```diff
@@ -213,7 +214,62 @@ impl MutblCap {
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+enum InheritedRefMatchRule {
+    /// Reference patterns try to consume the inherited reference first.
+    /// This assumes reference patterns can always match against an inherited reference.
```
That's exactly how I have it implemented locally, actually! I borrowed the field name from Typing Rust Patterns so it's even the same:
```rust
    EatBoth {
        /// Whether to allow matching on an inherited reference when matching against a
        /// non-reference type. This is `false` for current stable Rust.
        eat_inherited_ref_alone: bool,
    },
```
I agree the comments could use some work to be more clear about that though, yeah (that one included, on second glance.. "matching on" and "matching against" in the same sentence isn't great). I'll try and incorporate the framing you suggested

dianne replied:
I've added this clarification to the doc comments. I wonder if additional examples might help too.. something like `[&&mut x]: &[&mut CT]` working in eat-outer vs. `[&mut &x]: &[&mut CT]` working in eat-inner. It's just a bit wordy.. especially when adapted to more believable types. It may be clearer without the extra bulk.

Nadrieril replied:
I like the changes you made, that feels enough for now.

dianne commented on compiler/rustc_hir_typeck/src/pat.rs:
```diff
@@ -213,7 +214,62 @@ impl MutblCap {
+            } else if self.tcx.features().ref_pat_eat_one_layer_2024_structural() {
+                InheritedRefMatchRule::EatInner
+            } else {
+                // Currently, matching against an inherited ref on edition 2024 is an error.
+                // Use `EatBoth` as a fallback to be similar to stable Rust.
+                InheritedRefMatchRule::EatBoth
```
I considered that, and I agree that'd be better for clarity, but I opted against it because I figured it'd be the exact same as `EatBoth { eat_inherited_ref_alone: false }`. Both rules would need to set up the pattern migration lint at least (in old editions for migration purposes and in 2024 for the error). And defaulting to `EatBoth` helps avoid errors other than those from the migration lint when migrating code from old editions to 2024. This is especially important since I think the migration lint (as currently implemented) won't be emitted unless we pass typeck. Though I don't think the lint needs to be emitted in THIR construction, so we could possibly remove this limitation by moving it.. I'd have to think about what sort of fallback behavior we'd want from an `InheritedRefMatchRule::Error` in that case.

Nadrieril replied:
Ah yeah we do want to pass typeck before we emit the lint I think, I wouldn't want to incorrectly emit it if there's actually a type error around. That's more of a diagnostic polishing issue, let's leave that for later.

dianne commented on compiler/rustc_hir_typeck/src/pat.rs:
```diff
@@ -2205,51 +2261,68 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                    } else {
+                        // The expected type isn't a reference, so match against the inherited ref.
+                        if pat_mutbl > inh_mut {
+                            // We can't match an inherited shared reference with `&mut`. This will
+                            // be a type error later, since we're matching a reference pattern
+                            // against a non-reference type.
```
I partially address this in the next PR, since the other ruleset had a diagnostic for that already: https://github.com/rust-lang/rust/pull/135434/commits/1931a6470bf17796dd1401a5df01fa827a5fdd56
I haven't updated the wording, formatting, etc., though. At the very least, it still needs an explanation/elaboration of the term "inherited", a link to reference material, etc.. [It will also need some explanation of fallback-to-outer, when relevant](https://github.com/rust-lang/rust/pull/135434/commits/ddf74a479527b0879eca05491bffcae1cd5fed30#r1919542187).

Nadrieril replied:
Let's leave diagnostic polishing to after t-lang has decided on the proposal they want.

dianne replied:
That's my thought also, yeah.

dianne committed:
Add comments explaining inherited references

This adds explanation of inherited references and how they relate to the default binding mode.

Co-authored-by: Nadrieril <Nadrieril@users.noreply.github.com>

dianne committed:
Add debug-assertions for assumptions about enabled typing rules

dianne committed:
elaborate/clarify the comments on `InheritedRefMatchRule`'s variants

dianne commented:
I've added your suggested comments and debug asserts; ty! I'll make a note to write some tests for mixed-edition patterns too.

@rustbot ready

Nadrieril commented:
Let's goo

<!--
@bors r+
-->

![bors-r-plus](https://www.pietroalbini.org/bors-r-plus.png)

bors commented:
:pushpin: Commit 0263772bac97b00c242bb13844a90e1a6cd652f3 has been approved by `Nadrieril`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Nadrieril 0263772bac97b00c242bb13844a90e1a6cd652f3 -->
<!-- homu: {"type":"Approved","sha":"0263772bac97b00c242bb13844a90e1a6cd652f3","approver":"Nadrieril","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
